internal/xds/rbac/matchers.go

```diff
@@ -361,26 +361,61 @@ func newPortMatcher(destinationPort uint32) *portMatcher {
 	return &portMatcher{destinationPort: destinationPort}
 }
 
-func (pm *portMatcher) match(data *RPCData) bool {
-	return data.DestinationPort == pm.destinationPort
+func (pm *portMatcher) match(data *rpcData) bool {
+	return data.destinationPort == pm.destinationPort
 }
 
 // authenticatedMatcher matches on the name of the Principal. If set, the URI
 // SAN or DNS SAN in that order is used from the certificate, otherwise the
 // subject field is used. If unset, it applies to any user that is
 // authenticated. authenticatedMatcher implements the matcher interface.
 type authenticatedMatcher struct {
-	stringMatcher internalmatcher.StringMatcher
+	stringMatcher *internalmatcher.StringMatcher
 }
 
 func newAuthenticatedMatcher(authenticatedMatcherConfig *v3rbacpb.Principal_Authenticated) (*authenticatedMatcher, error) {
+	// Represents this line in the RBAC documentation = "If unset, it applies to
+	// any user that is authenticated" (see package-level comments).
+	if authenticatedMatcherConfig.PrincipalName == nil {
+		return &authenticatedMatcher{}, nil
+	}
 	stringMatcher, err := internalmatcher.StringMatcherFromProto(authenticatedMatcherConfig.PrincipalName)
 	if err != nil {
 		return nil, err
 	}
-	return &authenticatedMatcher{stringMatcher: stringMatcher}, nil
+	return &authenticatedMatcher{stringMatcher: &stringMatcher}, nil
 }
 
-func (am *authenticatedMatcher) match(data *RPCData) bool {
-	return am.stringMatcher.Match(data.PrincipalName)
+func (am *authenticatedMatcher) match(data *rpcData) bool {
+	// Represents this line in the RBAC documentation = "If unset, it applies to
+	// any user that is authenticated" (see package-level comments). An
+	// authenticated downstream in a stateful TLS connection will have to
+	// provide a certificate to prove their identity. Thus, you can simply check
+	// if there is a certificate present.
+	if am.stringMatcher == nil {
+		return len(data.certs) != 0
+	}
+
+	// The order of matching as per the RBAC documentation (see package-level comments)
+	// is as follows: URI SANs, DNS SANs, and then subject name.
+	for _, cert := range data.certs {
```

Comment:
Ah, similar to grpc-java: (https://github.com/grpc/grpc-java/blob/c8ba60152958cb0f37d0e9a32ae406ce8d5f1ff0/xds/src/main/java/io/grpc/xds/internal/rbac/engine/GrpcAuthorizationEngine.java#L280). Thanks for pointing this out, and sorry for the delay in answering. Envoy seems to persist a singular cert, where as in grpc we persist a list. Rather than loop through all the certificates, I only checked the first certificate.